src/File.hack

```diff
@@ -26,34 +26,63 @@ final class File {
     return Str\split($this->contents, "\n");
   }
 
+  /** 1-based line number */
   const type TLineNumber = int;
 
   /**
-   * Returns the 1-based line number of lint markers in the file.
+   * The lint error code to suppress, which could be either a linter name or a
+   * lint rule ID.
+   */
+  const type TErrorCode = string;
+
+  /**
+   * Returns the error codes to suppress according to lint markers in the file.
    */
   <<__Memoize>>
-  public function lintMarkersForLineBasedSuppression(
-  ): dict<string, keyset<self::TLineNumber>> {
-    return $this->getLines()
-      |> Vec\map_with_key(
+  public function errorCodesToSuppress(): shape(
+    'error_codes_in_file_lint_markers' => keyset<this::TErrorCode>,
```

Comment:
maybe 'whole_file' and 'single_instance' as the keys? Seems both clearer, and less verbose/repetitive at callsites

There's no syntactic difference between 'this suppression applies to the next line' and 'this suppression applies to the next AST node'